src/ruis/render/scene/gltf_loader.cpp

```diff
@@ -0,0 +1,772 @@
+		mat.get().tex_arm = textures[arm_index];
+
+	return mat;
+}
+
+template <typename tp_type>
+std::vector<utki::shared_ref<tp_type>> gltf_loader::read_root_array(
+	std::function<tp_type(const jsondom::value& j)> read_func,
+	const jsondom::value& root_json,
+	const std::string& name
+)
+{
+	std::vector<utki::shared_ref<tp_type>> all;
+	// std::cout << "loading " << name << std::endl;
+	auto it = root_json.object().find(name);
+	if (it != root_json.object().end() && it->second.is_array()) {
+		for (const auto& sub_json : it->second.array()) {
+			all.push_back(read_func(sub_json));
+		}
+	}
+	return all;
+}
+
+utki::shared_ref<scene> gltf_loader::load(const papki::file& fi)
+{
+	auto gltf = fi.load();
+	utki::deserializer d(gltf);
+
+	constexpr auto gltf_header_size = 4;
+	if (auto header_sv = d.read_string(gltf_header_size); header_sv != "glTF"sv) {
+		throw std::invalid_argument(utki::cat("read_gltf(): file header is not 'glTF': ", header_sv));
+	}
+
+	auto version = d.read_uint32_le();
+	constexpr auto expected_gltf_version = 2;
+	if (version != expected_gltf_version) {
+		throw std::invalid_argument(utki::cat("read_gltf(): glTF file version is not as expected: ", version));
+	}
+
+	auto gltf_length = d.read_uint32_le();
+	if (gltf_length != gltf.size()) {
+		throw std::invalid_argument(
+			utki::cat("read_gltf(): glTF file size ", gltf.size(), " does not match length ", gltf_length)
+		);
+	}
+
+	auto chunk_length = d.read_uint32_le();
+	if (chunk_length == 0) {
+		throw std::invalid_argument("read_gltf(): chunk length = 0");
+	}
+	constexpr auto chunk_type_length = 4;
+	if (auto chunk_type = d.read_string(chunk_type_length); chunk_type != "JSON"sv) {
+		throw std::invalid_argument(
+			utki::cat("read_gltf(): unexpected first chunk type: ", chunk_type, ", expected JSON")
+		);
+	}
+
+	auto json_span = d.read_span(chunk_length);
+
+	chunk_length = d.read_uint32_le();
+	if (chunk_length == 0) {
+		throw std::invalid_argument("read_gltf(): chunk length = 0");
+	}
+	if (auto chunk_type = d.read_string(chunk_type_length); chunk_type != "BIN\0"sv) {
+		throw std::invalid_argument(
+			utki::cat("read_gltf(): unexpected first chunk type: ", chunk_type, ", expected JSON")
+		);
+	}
+
+	glb_binary_buffer = d.read_span(chunk_length);
+
+	auto json = jsondom::read(json_span);
+	ASSERT(json.is_object())
+
+	// for (const auto& kv : json.object()) {
+	// std::cout << "key = " << kv.first << ", value type = " << unsigned(kv.second.get_type()) << std::endl;
+	//}
+
+	// std::cout << "loading bufferViews" << std::endl;
+	{
+		auto it = json.object().find("bufferViews");
+		if (it == json.object().end() || !it->second.is_array()) {
+			throw std::invalid_argument("read_gltf(): glTF does not have any valid bufferViews");
+		} else {
+			for (const auto& buffer_view_json : it->second.array()) {
+				buffer_views.push_back(read_buffer_view(buffer_view_json));
+			}
+		}
+	}
+
+	//////////////////////////////////////////////////////////////////////
```

Comment:
what does this comment separator separate?

Reply:
removed